Recipe for a runtime without FRAME

JoshOrndorff opened the pull request:
This PR addresses #131 by demonstrating a runtime that does not use Frame. The runtime is as minimal as reasonable and has a single boolean storage value. Transaction can either set, clear, or toggle this storage value. Transaction are not signed.


- [x] Runtime compiles
- [x] Outer nodes compile with frameless runtime
- [ ] Writeup
- [x] Block production works
- [ ] Transactions can be submitted (example of [custom transaction encoding](https://github.com/paritytech/substrate-lfs/blob/master/demo/rpc-client/src/main.rs))
- [ ] Tests to ensure state transitions correctly. (Maybe completed in a followup)

JoshOrndorff committed:
Copy super runtime and strip frame pallets

JoshOrndorff committed:
Strip grandpa

JoshOrndorff committed:
Diving in. Gonna need help.

JoshOrndorff committed:
Something compiles

JoshOrndorff committed:
implemented execute_block

JoshOrndorff committed:
"implemented" initialize_block

JoshOrndorff committed:
implemented finalize_block

JoshOrndorff committed:
Implement Aura and Grandpa runtime APIs

JoshOrndorff committed:
Partial attempt at complete node. Not working. Comitting because pair-hack got cut short.

JoshOrndorff committed:
Solved a few warnings and trivial errors. Now really stuck.

gnunicorn committed:
remove frame references

gnunicorn committed:
fix unchecked extrinsic

gnunicorn committed:
the outer knows only about the opaque block

gnunicorn committed:
fix warnings

gnunicorn committed:
GenesisConfig defaultable

gnunicorn committed:
Implement so service builder can build

JoshOrndorff committed:
master.into()

JoshOrndorff committed:
Mostly working code, sketch of writeup

JoshOrndorff committed:
update Cargo.toml s to ba5c85a

JoshOrndorff committed:
Fix runtimes

JoshOrndorff committed:
Fix weights

JoshOrndorff committed:
Fix kitchen node

JoshOrndorff committed:
Fix rpc node

JoshOrndorff committed:
Update deps to 0bd9ffa

JoshOrndorff committed:
update compiler

JoshOrndorff committed:
Add dedicated pow runtime. basic-pow now builds and runs.

JoshOrndorff committed:
Clean code comments

JoshOrndorff committed:
strip down pow-runtime

JoshOrndorff committed:
Clean up genesis

JoshOrndorff committed:
Raise the difficulty

JoshOrndorff committed:
More write-up

JoshOrndorff committed:
simple -> basic

JoshOrndorff committed:
More writeup

JoshOrndorff committed:
Finish drafting writeup.

JoshOrndorff committed:
Merge branch 'master' into joshy-pow

JoshOrndorff committed:
link checker adjustment

JoshOrndorff committed:
Update nodes/basic-pow/src/pow.rs

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update runtimes/pow-runtime/src/lib.rs

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update book.toml

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update nodes/basic-pow/src/chain_spec.rs

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Update text/3-entrees/basic-pow.md

Co-Authored-By: joe petrowski <25483142+joepetrowski@users.noreply.github.com>

JoshOrndorff committed:
Clean some commented code

JoshOrndorff committed:
Bring simple-map up to snuff (grr this is mixed with a bigger PR, oh well)

JoshOrndorff committed:
Update all cargo files.

JoshOrndorff committed:
Update basic-token (except tests)

I'm not fixing the tests unless I learn that `MigrateAccount` type is 
sticking around long term.

JoshOrndorff committed:
Update struct storage

JoshOrndorff committed:
update execution-schedule

JoshOrndorff committed:
Ditch Linked Maps

JoshOrndorff committed:
Everything compiles (excet tests)

JoshOrndorff committed:
Simplify key and address handling in chain-spec

JoshOrndorff committed:
Merge branch 'joshy-pow' of github.com:substrate-developer-hub/recipes into joshy-pow

JoshOrndorff committed:
Update runtime toward apps compatability

JoshOrndorff committed:
Include this javascript diagnostic stuff temporarily.

JoshOrndorff committed:
Fix doc typo (Just so CI runs again)

JoshOrndorff committed:
Make pow-runtime code look as close as possible to api-runtime. (flailing)

JoshOrndorff committed:
Update to 34c5cb3. Everything compiles and tests pass.

JoshOrndorff committed:
Update to 1d8fa43 (all compiles)

JoshOrndorff committed:
Merge branch 'master' into follow-substrate-master

JoshOrndorff committed:
Merge branch 'joshy-pow' into follow-substrate-master

JoshOrndorff committed:
Merge branch 'master' into follow-substrate-master

JoshOrndorff committed:
restore Cargo.lock

JoshOrndorff committed:
Fix pow-runtime

JoshOrndorff committed:
Merge branch 'master' into follow-substrate-master

JoshOrndorff committed:
Merge branch 'master' into follow-substrate-master

JoshOrndorff committed:
de-alias super-runtime (finally)

JoshOrndorff committed:
Update deps to crates.io (alpha.5)

JoshOrndorff committed:
fix in-code build scripts alias

JoshOrndorff committed:
clean basic-pow cargo.toml

JoshOrndorff committed:
Clean Kitchen node cargo.toml

JoshOrndorff committed:
Include missing struct storage type

JoshOrndorff committed:
Clean rpc-node

JoshOrndorff committed:
Update version in Readme

JoshOrndorff committed:
clean default-instance

JoshOrndorff committed:
clean last-caller

JoshOrndorff committed:
Substrate is on crates.io Hooray!

JoshOrndorff committed:
Fix tests on new recipes

JoshOrndorff committed:
Merge master for new folder structure. No actual updates yet.

JoshOrndorff committed:
update wasm-build-runner

JoshOrndorff committed:
Update deps to alpha.5 (no compile)

JoshOrndorff committed:
Runtime Compiles

JoshOrndorff committed:
Dealias crates, general tidying

JoshOrndorff committed:
spaces to tabs

JoshOrndorff committed:
minor cleaning to node in case we end up keeping it

JoshOrndorff committed:
Merge branch 'alpha.5' into joshy-frameless-runtime

JoshOrndorff committed:
Compiles with a PoW node, but errors out immediately

JoshOrndorff committed:
Start updating frameless-node, but lots of API changes ahead

JoshOrndorff committed:
Dupe code from kitchen node

JoshOrndorff committed:
Seperate crate for sha3pow

JoshOrndorff committed:
Something compiles.

JoshOrndorff committed:
Create a new runtime so this thing actually runs.

JoshOrndorff committed:
Remove commented babe code

JoshOrndorff committed:
Wire existing basic-pow node to newly-separated sha3pow crate.

JoshOrndorff committed:
Bring in complete sha3pow from utxo workshop, and adapt nodes to use it.

JoshOrndorff committed:
Remove unnecessary trait bounds on Sha3Algorithm

JoshOrndorff committed:
Writeups for basic-pow and sha3-pow-consensus

JoshOrndorff committed:
Merge branch 'master' into joshy-hybrid-consensus

jimmychu0807 committed:
Minor fix & visible WIP note :)

JoshOrndorff committed:
update to custom Substrate branch to support justification import.

JoshOrndorff committed:
standardize super runtime TODO How to remove the session key stuff from the runtime?

JoshOrndorff committed:
Preserve babe-grandpa node and runtime. This will be a reipe about grabbing authorities fro mthe runtime.

JoshOrndorff committed:
standardize basic-pow node and prune pow-runtime

JoshOrndorff committed:
Standardize weight-fee-runtime.

JoshOrndorff committed:
Move sum-storage to super-runtime and prune api-runtime

JoshOrndorff committed:
Convert kitchen node to instant seal

JoshOrndorff committed:
Add comment with hint about light client problem

JoshOrndorff committed:
Merge branch 'master' into joshy-standardize-runtimes

JoshOrndorff committed:
Fix minor type aggregation issue

JoshOrndorff committed:
Convert rpc-node to instant seal

JoshOrndorff committed:
Only kitchen node supports ocw-runtime

JoshOrndorff committed:
Properly prune session keys from runtime

JoshOrndorff committed:
Standardize ocw-runtime

JoshOrndorff committed:
Restore api-runtime

JoshOrndorff committed:
Standardize API Runtime

JoshOrndorff committed:
rename sum-storage-runtime-api It isn't _necessarily_ related to the rpc

JoshOrndorff committed:
Update runtime API writeup to reflect refactorings.

JoshOrndorff committed:
Update nodes/babe-grandpa-node/Cargo.toml

Co-authored-by: David <dvdplm@gmail.com>

JoshOrndorff committed:
Update runtimes/super-runtime/src/lib.rs

Co-authored-by: David <dvdplm@gmail.com>

JoshOrndorff committed:
Update nodes/babe-grandpa-node/Cargo.toml

JoshOrndorff committed:
Kitchen node compiles with light client. Also converted api-runtime a sum-storage along the way.

JoshOrndorff committed:
Add fixme annotation to light clients.

JoshOrndorff committed:
Merge branch 'joshy-standardize-runtimes' of github.com:substrate-developer-hub/recipes into joshy-standardize-runtimes

JoshOrndorff committed:
Clean kitchen node Cargo.toml

JoshOrndorff committed:
Strip instant seal code from manual seal node

JoshOrndorff committed:
Update manual seal writeup

JoshOrndorff committed:
Write up kitchen node

JoshOrndorff committed:
Clarify language in api-runtime writeup

JoshOrndorff committed:
typo: because -> became

JoshOrndorff committed:
Replace TODO with comment in babe-grandpa node Cargo.toml

JoshOrndorff committed:
Remove some TODOs

JoshOrndorff committed:
bump deps, LongestChain, default weights

JoshOrndorff committed:
changed append api

JoshOrndorff committed:
API Runtime Compiles

JoshOrndorff committed:
Start working on OCW runtime and pallet

JoshOrndorff committed:
Super runtime compiles

JoshOrndorff committed:
Weight fee runtime and weight pallet compile. Both will eed another look before merging this.

JoshOrndorff committed:
Babe-grandpa-untime compiles

JoshOrndorff committed:
Merge branch 'master' into alpha.7

JoshOrndorff committed:
re-apply dependency bumps (not sure where that work went)

JoshOrndorff committed:
Babe-grandpa-node compiles

JoshOrndorff committed:
basic-pow compiles

jimmychu0807 committed:
updated for alpha.7

jimmychu0807 committed:
it compiles!

jimmychu0807 committed:
[wip] Fixing kitchen node compiler error

JoshOrndorff committed:
Upstream fix merged

JoshOrndorff committed:
Merge branch 'manual-seal-light-client' into alpha.7

JoshOrndorff committed:
minor fix to kitchen node text

JoshOrndorff committed:
bump dependencies and standardize format in all pallets' Cargo.toml files

JoshOrndorff committed:
Bump deps in all runtimes TODO Babe grandpa runtime doesn't compile

JoshOrndorff committed:
Make Babe Grandpa Runtime Compile

JoshOrndorff committed:
Bump deps in all nodes and make them compile

JoshOrndorff committed:
Start updating tests

JoshOrndorff committed:
fix charity tests

JoshOrndorff committed:
Most tests are updated

JoshOrndorff committed:
prune accidentally committed notes file

JoshOrndorff committed:
Start working on offchain-demo tests

JoshOrndorff committed:
Merge alpha.7 and start updating

JoshOrndorff committed:
Rename to optional-grandpa-runtime, update deps, make compile

JoshOrndorff committed:
update optional-grandpa genesis

JoshOrndorff committed:
rename back to minimal grandpa runtime

JoshOrndorff committed:
writeup hybrid consensus

JoshOrndorff committed:
hybrid consensus is no longer wip

jimmychu0807 committed:
offchain-demo tests pass

JoshOrndorff committed:
Merge branch 'master' into alpha.7

JoshOrndorff committed:
Merge branch 'develop' into joshy-frameless-runtime

JoshOrndorff committed:
Final nail in pow runtime

JoshOrndorff committed:
prune stray dependencies on safe-mix

JoshOrndorff committed:
Merge develop

JoshOrndorff committed:
Compiles with kitchen node

JoshOrndorff committed:
Prune frameless node

JoshOrndorff committed:
Prune some commented code

JoshOrndorff committed:
Sketch writeup

JoshOrndorff committed:
Try runtime in pow and manual seal Same Error

JoshOrndorff committed:
Insert some genesis state

JoshOrndorff committed:
Node doesn't crash

JoshOrndorff committed:
Merge branch 'joshy-prune-safe-mix' into develop

JoshOrndorff committed:
Remove no-longer-correct type override for `weight: u32`. (#246)

Also fixes copy-pasta to babe-grandpa-runtime types.

JoshOrndorff committed:
Merge branch 'develop' of github.com:substrate-developer-hub/recipes into develop

bkchr commented on runtimes/frameless-runtime/src/lib.rs:
```diff
@@ -0,0 +1,303 @@
+			VERSION
+		}
+
+		fn execute_block(block: Block) {
```
This function is wrong. You don't call `intialize_block` and `finalize_block`.

JoshOrndorff replied:
I've added a commit which calls those functions, but the behavior is the same as before. Maybe I called them incorrectly?

bkchr replied:
Your `apply_extrinsic` function is not doing anything. And your `execute_block` function handles the transactions. I'm not sure how you expect this to work? You need to do the exactly same operations while building and importing the block.

JoshOrndorff committed:
Call on_initialize and on_finalize from execute_block

JoshOrndorff committed:
Add image in ocw storage recipe

commit 98a974fd12cc72a07843fd927b4abdfd68baf444
Author: Jimmy Chu <jimmychu0807@gmail.com>
Date:   Fri May 15 14:44:11 2020 +0800

    fix: adding image in ocw storage recipe

commit a4ccaed116b8e1e44d9af530375f82c4987fdbfd
Author: Joshy Orndorff <JoshOrndorff@users.noreply.github.com>
Date:   Thu May 14 15:35:23 2020 -0400

    prune stray dependencies on safe-mix (#245)

JoshOrndorff committed:
Update CI to support gitflow (#247)

* Modify CI to support gitflow

* fix syntax error

JoshOrndorff committed:
Update to alpha.8 (#249)

* bump deps

* CheckVersion -> CheckTxVersion

* Make child tree copmile.

TODO revise child trie and simple crowdfund in the future.

* fix ocw-runtime

* registry

JoshOrndorff committed:
add .md extension (#251)

JoshOrndorff committed:
cargo fmt (#253)

* config file to preserve tabs

* cargo fmt. Do we like all these reformats?

I think mostly yes, but let's evaluate carefully, and maybe update the 
config

JoshOrndorff committed:
Update Contributing Guidelines (#252)

* Revise contributing docs

* Update Readme.

* add link to issue queue

* Update CONTRIBUTING.md

Co-authored-by: Jimmy Chu <jimmy@parity.io>

* Update CONTRIBUTING.md

Co-authored-by: Jimmy Chu <jimmy@parity.io>

* Update CONTRIBUTING.md

* Update README.md

Co-authored-by: Jimmy Chu <jimmy@parity.io>

JoshOrndorff committed:
Add clippy to CI (#250)

* Add clippy to CI

* Try running clippy wih `+nightly`

* Start linting babe grandpa node

* start linting basic-pow

* Add explicit target

* lint adding machien

* lint basic-token

* lint c and d pallets

* lint e-o pallets

* lint remaining pallets

* Suppress clippy warning on babe-grandpa runtime

* Suppress clippy warning in min grandpa runtime

* lint hybrid-consensus

* lint kitchen node

* lint manual seal

* Revert "Add explicit target"

This reverts commit a476eb59b221b7ee4cd8cdeba134a004151bceb0.

* Try with override

* Update .github/workflows/test-code.yml

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* one more lint in fixed-point

* lint rpc-node

* Suppress warning present only in wasm.

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

JoshOrndorff committed:
Clean cargo.toml files (#254)

* default dash features

* dealias a few system vs fram-system imports

JoshOrndorff committed:
Revise prepare kitchen chapter (#255)

* Explain instant seal

* Briefly mention consensus directory

* Update quoted code from kitchen-node/Cargo.toml

JoshOrndorff committed:
bump dependencies

JoshOrndorff committed:
ProposerFactory takes prometheus registry

JoshOrndorff committed:
No longer need to allow clippy::string_lit_as_bytes

JoshOrndorff committed:
Substrate dependencies to 12e08fd (#257)

* bump dependencies

* ProposerFactory takes prometheus registry

* No longer need to allow clippy::string_lit_as_bytes

JoshOrndorff committed:
update rustdocs links

JoshOrndorff committed:
add rustdocs update to release process

JoshOrndorff committed:
revert commits accidentally made directly to develop

JoshOrndorff committed:
Change rustdocs link to crates.parity.io (#260)

* Revert the revert, thus restoring the change

This reverts commit 0e4fbfd0cb81c91ac549d6b4a04036393c25850a.

* fix link in weights

* fix links to client

* pull in jimmy's changes initially proposed against master.

JoshOrndorff committed:
Merge branch 'develop' into prettier-test

JoshOrndorff committed:
fix clippy error

JoshOrndorff committed:
Format markdown with Prettier (#263)

* add config

* run on markdown files

JoshOrndorff committed:
fix Summary

JoshOrndorff committed:
Cleanup crowdfund (#265)

* simple crowdfund compiles once again

* cargo fmt

* decl_error and other fixes

* clean comments and spacing

* Start scaffolding tests

* Dispense and more tests

* More tests

* More tests and cargo fmt

* Install in super runtime

* cleanup types

* writeup

* prettier

* Update Summary

JoshOrndorff committed:
Add writeup for Babe Grandpa Node (#264)

* strip ocw stuff from babe-grandpa

* Init babe grandpa writeup

* Refactr babe-grandpa node writeup to reference KB for import pipeline

* insert link to kb import pipeline from basic-pow writeup

* link to kb block import from hybrid consensus writeup

* rustdocs links to crates.parity.io

* fix more links

* Remove accidentally-duplicated TOC items

* Fix caps

* Rephrase compatability issue

JoshOrndorff committed:
Prepare Cargo.toml files for publishing (#226)

* Update Cargo.toml for publishing

* Update cargo.toml for publishing tom arketplace

* update more cargo.toml files

* Finish updating pallet Cargo toml files

* update runtime cargotoml files

* update some missed -dev s

* Minor update to storage api index.

lsaether committed:
Fix typos in kitchen-node doc (#268)

Noticed some typos while reading.

JoshOrndorff committed:
Fix name on kitchen node write-up

JoshOrndorff committed:
Update dependencies to 45b9f0a (rc2) (#269)

* bump deps

* Super Runtime Compiles

* Most runtimes compile

* Weight fee runtime compiles and updated for new API

* Update fees writeup

* Fix typos in key names

* fix test harnesses

* Remove dupe type in charity tests

* Update text/3-entrees/fees.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/fees.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/fees.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/fees.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/fees.md

Co-authored-by: Dan Forbes <dan@parity.io>

* link to pallet transaction payment

Co-authored-by: Dan Forbes <dan@parity.io>

JoshOrndorff committed:
Merge branch 'develop' of github.com:substrate-developer-hub/recipes into develop

JoshOrndorff committed:
Bring changes from rc2-prep back to develop branch (#274)

* prune stray dependencies on safe-mix (#245)

* fix: adding image in ocw storage recipe (#248)

* Release alpha.8 (#256)

* Dupe code from kitchen node

* Seperate crate for sha3pow

* Something compiles.

* Create a new runtime so this thing actually runs.

* Remove commented babe code

* Wire existing basic-pow node to newly-separated sha3pow crate.

* Bring in complete sha3pow from utxo workshop, and adapt nodes to use it.

* Remove unnecessary trait bounds on Sha3Algorithm

* Writeups for basic-pow and sha3-pow-consensus

* Minor fix & visible WIP note :)

* update to custom Substrate branch to support justification import.

* standardize super runtime TODO How to remove the session key stuff from the runtime?

* Preserve babe-grandpa node and runtime. This will be a reipe about grabbing authorities fro mthe runtime.

* standardize basic-pow node and prune pow-runtime

* Standardize weight-fee-runtime.

* Move sum-storage to super-runtime and prune api-runtime

* Convert kitchen node to instant seal

* Add comment with hint about light client problem

* Fix minor type aggregation issue

* Convert rpc-node to instant seal

* Only kitchen node supports ocw-runtime

* Properly prune session keys from runtime

* Standardize ocw-runtime

* Restore api-runtime

* Standardize API Runtime

* rename sum-storage-runtime-api It isn't _necessarily_ related to the rpc

* Update runtime API writeup to reflect refactorings.

* Update nodes/babe-grandpa-node/Cargo.toml

Co-authored-by: David <dvdplm@gmail.com>

* Update runtimes/super-runtime/src/lib.rs

Co-authored-by: David <dvdplm@gmail.com>

* Update nodes/babe-grandpa-node/Cargo.toml

* Kitchen node compiles with light client. Also converted api-runtime a sum-storage along the way.

* Add fixme annotation to light clients.

* Clean kitchen node Cargo.toml

* Strip instant seal code from manual seal node

* Update manual seal writeup

* Write up kitchen node

* Clarify language in api-runtime writeup

* typo: because -> became

* Replace TODO with comment in babe-grandpa node Cargo.toml

* Remove some TODOs

* bump deps, LongestChain, default weights

* changed append api

* API Runtime Compiles

* Start working on OCW runtime and pallet

* Super runtime compiles

* Weight fee runtime and weight pallet compile. Both will eed another look before merging this.

* Babe-grandpa-untime compiles

* re-apply dependency bumps (not sure where that work went)

* Babe-grandpa-node compiles

* basic-pow compiles

* updated for alpha.7

* it compiles!

* [wip] Fixing kitchen node compiler error

* Upstream fix merged

* minor fix to kitchen node text

* bump dependencies and standardize format in all pallets' Cargo.toml files

* Bump deps in all runtimes TODO Babe grandpa runtime doesn't compile

* Make Babe Grandpa Runtime Compile

* Bump deps in all nodes and make them compile

* Start updating tests

* fix charity tests

* Most tests are updated

* prune accidentally committed notes file

* Start working on offchain-demo tests

* Rename to optional-grandpa-runtime, update deps, make compile

* update optional-grandpa genesis

* rename back to minimal grandpa runtime

* writeup hybrid consensus

* hybrid consensus is no longer wip

* offchain-demo tests pass

* prune stray dependencies on safe-mix

* Remove no-longer-correct type override for `weight: u32`. (#246)

Also fixes copy-pasta to babe-grandpa-runtime types.

* Add image in ocw storage recipe

commit 98a974fd12cc72a07843fd927b4abdfd68baf444
Author: Jimmy Chu <jimmychu0807@gmail.com>
Date:   Fri May 15 14:44:11 2020 +0800

    fix: adding image in ocw storage recipe

commit a4ccaed116b8e1e44d9af530375f82c4987fdbfd
Author: Joshy Orndorff <JoshOrndorff@users.noreply.github.com>
Date:   Thu May 14 15:35:23 2020 -0400

    prune stray dependencies on safe-mix (#245)

* Update CI to support gitflow (#247)

* Modify CI to support gitflow

* fix syntax error

* Update to alpha.8 (#249)

* bump deps

* CheckVersion -> CheckTxVersion

* Make child tree copmile.

TODO revise child trie and simple crowdfund in the future.

* fix ocw-runtime

* registry

* add .md extension (#251)

* cargo fmt (#253)

* config file to preserve tabs

* cargo fmt. Do we like all these reformats?

I think mostly yes, but let's evaluate carefully, and maybe update the 
config

* Update Contributing Guidelines (#252)

* Revise contributing docs

* Update Readme.

* add link to issue queue

* Update CONTRIBUTING.md

Co-authored-by: Jimmy Chu <jimmy@parity.io>

* Update CONTRIBUTING.md

Co-authored-by: Jimmy Chu <jimmy@parity.io>

* Update CONTRIBUTING.md

* Update README.md

Co-authored-by: Jimmy Chu <jimmy@parity.io>

* Add clippy to CI (#250)

* Add clippy to CI

* Try running clippy wih `+nightly`

* Start linting babe grandpa node

* start linting basic-pow

* Add explicit target

* lint adding machien

* lint basic-token

* lint c and d pallets

* lint e-o pallets

* lint remaining pallets

* Suppress clippy warning on babe-grandpa runtime

* Suppress clippy warning in min grandpa runtime

* lint hybrid-consensus

* lint kitchen node

* lint manual seal

* Revert "Add explicit target"

This reverts commit a476eb59b221b7ee4cd8cdeba134a004151bceb0.

* Try with override

* Update .github/workflows/test-code.yml

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* one more lint in fixed-point

* lint rpc-node

* Suppress warning present only in wasm.

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* Clean cargo.toml files (#254)

* default dash features

* dealias a few system vs fram-system imports

* Revise prepare kitchen chapter (#255)

* Explain instant seal

* Briefly mention consensus directory

* Update quoted code from kitchen-node/Cargo.toml

* bump deps

* bump our own crate versions

Co-authored-by: Jimmy Chu <jimmychu0807@gmail.com>
Co-authored-by: David <dvdplm@gmail.com>
Co-authored-by: Jimmy Chu <jimmy@parity.io>
Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* Fix rustdocs links to alpha.8 version (#259)

* fix rustdocs links to alpha.8 version

* fix link in weights

* fix links to client

* remove broken ocw-related links

* Fix typos in kitchen-node doc (#268)

Noticed some typos while reading.

* bump deps in runtime

* bump deps in outer node

* bump our own versions

* update rustdocs links

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/babe-grandpa-node.md

* Update text/3-entrees/basic-pow.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/basic-pow.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update pallets/fixed-point/Cargo.toml

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/1-hello-substrate.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/1-hello-substrate.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/safemath.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/storage-api/sets-vecs-iteration.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/storage-api/ringbuffer.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/1-hello-substrate.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/2-storage-values.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/2-storage-values.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/2-storage-values.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/3-errors.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/3-errors.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/3-errors.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/3-errors.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Lots of nitpicks in text

Co-authored-by: David <dvdplm@gmail.com>

* Lots more nitpicks in text

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/testing/off-chain-workers.md

Co-authored-by: David <dvdplm@gmail.com>

* More nits fro mreview

Co-authored-by: David <dvdplm@gmail.com>

* Release RC2 (#271)

* Dupe code from kitchen node

* Seperate crate for sha3pow

* Something compiles.

* Create a new runtime so this thing actually runs.

* Remove commented babe code

* Wire existing basic-pow node to newly-separated sha3pow crate.

* Bring in complete sha3pow from utxo workshop, and adapt nodes to use it.

* Remove unnecessary trait bounds on Sha3Algorithm

* Writeups for basic-pow and sha3-pow-consensus

* Minor fix & visible WIP note :)

* update to custom Substrate branch to support justification import.

* standardize super runtime TODO How to remove the session key stuff from the runtime?

* Preserve babe-grandpa node and runtime. This will be a reipe about grabbing authorities fro mthe runtime.

* standardize basic-pow node and prune pow-runtime

* Standardize weight-fee-runtime.

* Move sum-storage to super-runtime and prune api-runtime

* Convert kitchen node to instant seal

* Add comment with hint about light client problem

* Fix minor type aggregation issue

* Convert rpc-node to instant seal

* Only kitchen node supports ocw-runtime

* Properly prune session keys from runtime

* Standardize ocw-runtime

* Restore api-runtime

* Standardize API Runtime

* rename sum-storage-runtime-api It isn't _necessarily_ related to the rpc

* Update runtime API writeup to reflect refactorings.

* Update nodes/babe-grandpa-node/Cargo.toml

Co-authored-by: David <dvdplm@gmail.com>

* Update runtimes/super-runtime/src/lib.rs

Co-authored-by: David <dvdplm@gmail.com>

* Update nodes/babe-grandpa-node/Cargo.toml

* Kitchen node compiles with light client. Also converted api-runtime a sum-storage along the way.

* Add fixme annotation to light clients.

* Clean kitchen node Cargo.toml

* Strip instant seal code from manual seal node

* Update manual seal writeup

* Write up kitchen node

* Clarify language in api-runtime writeup

* typo: because -> became

* Replace TODO with comment in babe-grandpa node Cargo.toml

* Remove some TODOs

* bump deps, LongestChain, default weights

* changed append api

* API Runtime Compiles

* Start working on OCW runtime and pallet

* Super runtime compiles

* Weight fee runtime and weight pallet compile. Both will eed another look before merging this.

* Babe-grandpa-untime compiles

* re-apply dependency bumps (not sure where that work went)

* Babe-grandpa-node compiles

* basic-pow compiles

* updated for alpha.7

* it compiles!

* [wip] Fixing kitchen node compiler error

* Upstream fix merged

* minor fix to kitchen node text

* bump dependencies and standardize format in all pallets' Cargo.toml files

* Bump deps in all runtimes TODO Babe grandpa runtime doesn't compile

* Make Babe Grandpa Runtime Compile

* Bump deps in all nodes and make them compile

* Start updating tests

* fix charity tests

* Most tests are updated

* prune accidentally committed notes file

* Start working on offchain-demo tests

* Rename to optional-grandpa-runtime, update deps, make compile

* update optional-grandpa genesis

* rename back to minimal grandpa runtime

* writeup hybrid consensus

* hybrid consensus is no longer wip

* offchain-demo tests pass

* prune stray dependencies on safe-mix

* Remove no-longer-correct type override for `weight: u32`. (#246)

Also fixes copy-pasta to babe-grandpa-runtime types.

* Add image in ocw storage recipe

commit 98a974fd12cc72a07843fd927b4abdfd68baf444
Author: Jimmy Chu <jimmychu0807@gmail.com>
Date:   Fri May 15 14:44:11 2020 +0800

    fix: adding image in ocw storage recipe

commit a4ccaed116b8e1e44d9af530375f82c4987fdbfd
Author: Joshy Orndorff <JoshOrndorff@users.noreply.github.com>
Date:   Thu May 14 15:35:23 2020 -0400

    prune stray dependencies on safe-mix (#245)

* Update CI to support gitflow (#247)

* Modify CI to support gitflow

* fix syntax error

* Update to alpha.8 (#249)

* bump deps

* CheckVersion -> CheckTxVersion

* Make child tree copmile.

TODO revise child trie and simple crowdfund in the future.

* fix ocw-runtime

* registry

* add .md extension (#251)

* cargo fmt (#253)

* config file to preserve tabs

* cargo fmt. Do we like all these reformats?

I think mostly yes, but let's evaluate carefully, and maybe update the 
config

* Update Contributing Guidelines (#252)

* Revise contributing docs

* Update Readme.

* add link to issue queue

* Update CONTRIBUTING.md

Co-authored-by: Jimmy Chu <jimmy@parity.io>

* Update CONTRIBUTING.md

Co-authored-by: Jimmy Chu <jimmy@parity.io>

* Update CONTRIBUTING.md

* Update README.md

Co-authored-by: Jimmy Chu <jimmy@parity.io>

* Add clippy to CI (#250)

* Add clippy to CI

* Try running clippy wih `+nightly`

* Start linting babe grandpa node

* start linting basic-pow

* Add explicit target

* lint adding machien

* lint basic-token

* lint c and d pallets

* lint e-o pallets

* lint remaining pallets

* Suppress clippy warning on babe-grandpa runtime

* Suppress clippy warning in min grandpa runtime

* lint hybrid-consensus

* lint kitchen node

* lint manual seal

* Revert "Add explicit target"

This reverts commit a476eb59b221b7ee4cd8cdeba134a004151bceb0.

* Try with override

* Update .github/workflows/test-code.yml

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* one more lint in fixed-point

* lint rpc-node

* Suppress warning present only in wasm.

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

* Clean cargo.toml files (#254)

* default dash features

* dealias a few system vs fram-system imports

* Revise prepare kitchen chapter (#255)

* Explain instant seal

* Briefly mention consensus directory

* Update quoted code from kitchen-node/Cargo.toml

* bump dependencies

* ProposerFactory takes prometheus registry

* No longer need to allow clippy::string_lit_as_bytes

* Substrate dependencies to 12e08fd (#257)

* bump dependencies

* ProposerFactory takes prometheus registry

* No longer need to allow clippy::string_lit_as_bytes

* update rustdocs links

* add rustdocs update to release process

* revert commits accidentally made directly to develop

* Change rustdocs link to crates.parity.io (#260)

* Revert the revert, thus restoring the change

This reverts commit 0e4fbfd0cb81c91ac549d6b4a04036393c25850a.

* fix link in weights

* fix links to client

* pull in jimmy's changes initially proposed against master.

* fix clippy error

* Format markdown with Prettier (#263)

* add config

* run on markdown files

* fix Summary

* Cleanup crowdfund (#265)

* simple crowdfund compiles once again

* cargo fmt

* decl_error and other fixes

* clean comments and spacing

* Start scaffolding tests

* Dispense and more tests

* More tests

* More tests and cargo fmt

* Install in super runtime

* cleanup types

* writeup

* prettier

* Update Summary

* Add writeup for Babe Grandpa Node (#264)

* strip ocw stuff from babe-grandpa

* Init babe grandpa writeup

* Refactr babe-grandpa node writeup to reference KB for import pipeline

* insert link to kb import pipeline from basic-pow writeup

* link to kb block import from hybrid consensus writeup

* rustdocs links to crates.parity.io

* fix more links

* Remove accidentally-duplicated TOC items

* Fix caps

* Rephrase compatability issue

* Prepare Cargo.toml files for publishing (#226)

* Update Cargo.toml for publishing

* Update cargo.toml for publishing tom arketplace

* update more cargo.toml files

* Finish updating pallet Cargo toml files

* update runtime cargotoml files

* update some missed -dev s

* Minor update to storage api index.

* Fix typos in kitchen-node doc (#268)

Noticed some typos while reading.

* Fix name on kitchen node write-up

* Update dependencies to 45b9f0a (rc2) (#269)

* bump deps

* Super Runtime Compiles

* Most runtimes compile

* Weight fee runtime compiles and updated for new API

* Update fees writeup

* Fix typos in key names

* fix test harnesses

* Remove dupe type in charity tests

* Update text/3-entrees/fees.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/fees.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/fees.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/fees.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/fees.md

Co-authored-by: Dan Forbes <dan@parity.io>

* link to pallet transaction payment

Co-authored-by: Dan Forbes <dan@parity.io>

* bump deps in runtime

* bump deps in outer node

* bump our own versions

* update rustdocs links

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/babe-grandpa-node.md

* Update text/3-entrees/basic-pow.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/basic-pow.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update pallets/fixed-point/Cargo.toml

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/1-hello-substrate.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/1-hello-substrate.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/safemath.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/storage-api/sets-vecs-iteration.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/storage-api/ringbuffer.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/1-hello-substrate.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/2-storage-values.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/2-storage-values.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/2-storage-values.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/3-errors.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/3-errors.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/3-errors.md

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/3-errors.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Lots of nitpicks in text

Co-authored-by: David <dvdplm@gmail.com>

* Lots more nitpicks in text

Co-authored-by: David <dvdplm@gmail.com>

* Update text/3-entrees/testing/off-chain-workers.md

Co-authored-by: David <dvdplm@gmail.com>

* More nits fro mreview

Co-authored-by: David <dvdplm@gmail.com>

Co-authored-by: Jimmy Chu <jimmychu0807@gmail.com>
Co-authored-by: David <dvdplm@gmail.com>
Co-authored-by: Jimmy Chu <jimmy@parity.io>
Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>
Co-authored-by: Logan Saether <lsaether@protonmail.com>
Co-authored-by: Dan Forbes <dan@parity.io>

* Revert release-prep steps

This reverts the forumlaic stuff that was done in preparation for 
releasing rc2, but not the actual fixes.
Concretely, this restoreds dependencies to github, links to 
crates.parity.io, and versions to 2.0.0-dev

* Update CONTRIBUTING.md

Co-authored-by: David <dvdplm@gmail.com>

* Update runtimes/ocw-runtime/src/lib.rs

Co-authored-by: David <dvdplm@gmail.com>

* Update text/2-appetizers/1-hello-substrate.md

Co-authored-by: David <dvdplm@gmail.com>

* Update pallets/fixed-point/Cargo.toml

Co-authored-by: Dan Forbes <dan@parity.io>

* Update runtimes/ocw-runtime/src/lib.rs

* Update CONTRIBUTING.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/2-appetizers/2-storage-values.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/basic-pow.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/crowdfund.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/storage-api/cache.md

Co-authored-by: Dan Forbes <dan@parity.io>

* Update text/3-entrees/storage-api/sets-vecs-iteration.md

Co-authored-by: Dan Forbes <dan@parity.io>

Co-authored-by: Jimmy Chu <jimmy@parity.io>
Co-authored-by: Jimmy Chu <jimmychu0807@gmail.com>
Co-authored-by: David <dvdplm@gmail.com>
Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>
Co-authored-by: Logan Saether <lsaether@protonmail.com>
Co-authored-by: Dan Forbes <dan@parity.io>

JoshOrndorff committed:
Update to cd305af (#275)

* bump dependencies

* fix runtimes. MaxExtrinsicWeight is not const

* cargo fmt

* prettier

* unfuck SUMMARY (prettier always does this)

JoshOrndorff committed:
Recipe for coupled pallets (#272)

* Sketch unifying trait

* update vec-set code

* impl AccountSet for vec-set

* sketch out map-set

* impl AccountSet for map-set

* update check-membership code

* write loosely coupled variant of check-membership

* tests for vec-set

* tests for map-set

* tests for membership check tight variant

* tests for moembership check loose variant

* Try to update super-runtime, but notice build is broken

* runtime compiles

* write up check membership

* prune stale draft, saving note about coupling to system pallet

* Merge branch 'develop' into coupled-pallets

* add max member count to map-set

* satisfy clippy

* Write up vec-set

* Write up map-set

* Fix nits found during code review

Co-authored-by: Dan Forbes <dan@parity.io>

* Update pallets/check-membership/src/lib.rs

Co-authored-by: Dan Forbes <dan@parity.io>

* One

Co-authored-by: David <dvdplm@gmail.com>

* File

Co-authored-by: David <dvdplm@gmail.com>

* At

Co-authored-by: David <dvdplm@gmail.com>

* A

Co-authored-by: Dan Forbes <dan@parity.io>
Co-authored-by: David <dvdplm@gmail.com>

* Time

Co-authored-by: David <dvdplm@gmail.com>
Co-authored-by: Dan Forbes <dan@parity.io>

* Last of the github suggestions

Co-authored-by: David <dvdplm@gmail.com>
Co-authored-by: Dan Forbes <dan@parity.io>

* Add example of loose coupling in FRAME

* fix mdbook build

* One last pass with prettier

Co-authored-by: Dan Forbes <dan@parity.io>
Co-authored-by: David <dvdplm@gmail.com>

JoshOrndorff committed:
Rename vec-set's store to VecSet (#279)

Commit:
Fix substrate.io links (#286)

* Replace substrate.io links

* Fix more broken links

JoshOrndorff committed:
Update CI Actions (#285)

* update checkout action

* update mdbook

* update cache action (merge 3 caches into 1)

manxiaqu committed:
fix typo and format style (#280)

TovarishFin committed:
Update lib.rs

fix small typos

JoshOrndorff committed:
Decouple link checking from mdbook (#287)

* Add new linkcheck ci

* strip old link checker

* Remove broken and outdated link

* Don't check links to github code.

These links will always be broken on PRs because the code in question is 
not yet merged. I hope t oreplae these links with playground integration 
soon anyway.

* remove crate.io link because they always come back 404 in linkcheck

* Fix realative link in ocw writeup

* fix more ocw links

* another ocw relative link

* html -> md to fix link

* don't check link that actually works but appears broken to checker :shrug:

* prune fuzzing draft that was had little content and a broken link

* prune maps draft

* clean links in optimization draft

* work around broken link in potentially useful draft

* fix draft link

* prune declarative syntax draft

* html -> md

* clean readme links

* try quiet mode

* WIP

* WIP

* html -> md

* fix ocw link

* try enabling config file

* disable config file

* try quiet mode again

* try again with config file

* fix typo in config file (I dont expect this will solve the issue with false positives)

* Revert addition on individual ignore comments

This reverts commit 9bb02fc65cb99a70b0c865c5c5403f8d66cb87fb.

* https for ignore patterns

* try other path to config file

* Introduce a broken link just to be double sure

* Revert "Introduce a broken link just to be double sure"

This reverts commit 80923e293d3eff06863ddbad29d25f09c05b4e38.

lsaether committed:
Addition to Manual Seal that combines it with Instant seal (#291)

* Get the basics down

* Add text

* Fix formatting

* Clean up

* Some wording fixes

igorline committed:
Fix typo (#292)

* Fix typo

* Fix example for 4-digit numbers multiplication

JoshOrndorff committed:
Updates Substrate dependencies to  34695a8 (#293)

* bump dependencies

* update transaction_pool api

JoshOrndorff committed:
develop.into()

JoshOrndorff committed:
Update license and other Cargo.toml metadata

JoshOrndorff committed:
x

JoshOrndorff committed:
Add debug prints and body for `apply_extrinsic`

JoshOrndorff committed:
More Cargo.toml cleaning

bkchr commented on runtimes/frameless-runtime/src/lib.rs:
```diff
@@ -0,0 +1,340 @@
+		Verify,
+	}
+};
+//TODO this is a weird import. Is this correct?
```
Yes. That is normally done by the macro.

bkchr replied:
But why do you need it?

JoshOrndorff replied:
I need it in order to call `Self::finalize_block()` from with `execute_block`.

bkchr commented on runtimes/frameless-runtime/src/lib.rs:
```diff
@@ -0,0 +1,340 @@
+
+			let raw_state_root = &sp_io::storage::root()[..];
+
+			let state_root = sp_core::H256::from(sp_io::hashing::blake2_256(raw_state_root));
```
Okay, how should this work? Did you tried to follow what you are writing to the storage?

The problem is, when you build the block, you generate the state root based on a header that has `0x00...` as `state_root` and when you call this function as part of `execute_block`, the header has a "real" state root. This leads to a different state root being generated.

JoshOrndorff replied:
Okay, after enough studying, I understand what you mean. I've changed the logic to clear the temporary-stored header out of storage before calculating the root. Now I get the same storage root when building and executing. I think this is a step in the right direction. However, the node still crashes with

```
Block prepare storage changes error:
InvalidStateRoot
```

JoshOrndorff committed:
Clear header from storage and log calculated header.

JoshOrndorff committed:
clippy

JoshOrndorff committed:
more clippy

bkchr commented on runtimes/frameless-runtime/src/lib.rs:
```diff
@@ -0,0 +1,341 @@
+
+			let raw_state_root = &sp_io::storage::root()[..];
+
+			let state_root = sp_core::H256::from(sp_io::hashing::blake2_256(raw_state_root));
```
This is not correct. not sure why you hash the state root here?
```suggestion
			let state_root = sp_core::H256::decode(&mut &raw_state_root[..]).unwrap();
```

JoshOrndorff committed:
update all authors (#296)

JoshOrndorff committed:
Don't manually hash state root

Co-authored-by: Bastian Köcher <bkchr@users.noreply.github.com>

JoshOrndorff committed:
Merge branch 'develop' into joshy-frameless-runtime

JoshOrndorff committed:
prune GetNodeBlock

JoshOrndorff committed:
Clean comments and style

JoshOrndorff committed:
strip commented code

JoshOrndorff committed:
Hack ValidTransaction

JoshOrndorff commented on runtimes/frameless-runtime/src/lib.rs:
```diff
@@ -0,0 +1,312 @@
+			_source: TransactionSource,
+			_tx: <Block as BlockT>::Extrinsic) -> TransactionValidity {
+			// Any transaction of the correct type is valid
+			Ok(ValidTransaction{
+				priority: 1u64,
+				requires: Vec::new(),
+				// This hach was necessary to make the node accept any transactions at all
+				// When I was setting provides to an empty vec, submiting a transaction failed
+				// the RPC responded {"code":-32603,"message":"Unknown error occurred","data":"Pool(NoTagsProvided)"}
+				// Adding this provides tag solved that. Solutions moving forward:
+				// 1. Require a nonce with each transaction
+				// 2. Try to relax the TxPool's requirement that every transaction provide some tag
+				provides: vec![vec![0]],
+				longevity: TransactionLongevity::max_value(),
+				propagate: true,
+			})
```
@todr The intention of this simple runtime is that transactions can be applied in absolutely any order. There are no dependencies between transactions and no meaningful tags to associate with the transactions. But https://github.com/paritytech/substrate/blob/75113aae02ee3cc265aaaa13cdf5a79c7ff98105/client/transaction-pool/graph/src/pool.rs#L410-L412 enforces that each transaction provide at least one tag. Is there any harm in removing this requirement?

JoshOrndorff replied:
Oops... @tomusdrw

tomusdrw replied:
Sorry, somehow missed the previous mention.

The dependencies are indicated via `requires` tag set, `provides` guarantees de-duplication of the transactions.
Imagine that we allow no `provides` tag. The same transaction can then be included in the pool and in the block as many times as you would like - so each time you receive the transaction over network from your peers it would be imported to the pool and added to the block, so you would end up with infinite number of copies of the same transaction.

I'd suggest adding `provides: vec![hash(tx.encode())]` here. This will de-duplicate transactions floating around the network, but note that it doesn't prevent replaying the same transaction in consecutive blocks.
The runtime must include some kind of replay protection and at some point return `Stale` from `validate_transaction` for transactions that have already been applied in the network.

I'd suggest adding at least `apply_before_block` parameter to each extrinsic, so that we can detect stale transactions. Note that it still does not prevent replaying the same transaction over and over again (in the same block or subsequent blocks) when it's valid.

tomusdrw replied:
A scheme I had in mind to allow both sequential and parallel transactions some time ago looks like this.
1. Each transaction has `(random_salt, nonce)`.
2. If you want transactions to be included in parallel you set `nonce=0` and choose a random `salt`.
3. If you want transactions to be included sequentially you use the same `salt` and increment the `nonce`.
4. Runtime maintains storage entries for:
```rust
nonce_expirations: DoubleMap<BlockNumber, Salt, ()>,
next_nonce: Map<Salt, (Nonce, BlockNumber)>,
```
Salts are valid for fixed `X` blocks, and every time a salt is used we do something like this:
```
let expire_at = System::current_block_number() + X;
// Update next expected nonce and pruning block.
let (nonce, block) = next_nonce.get(salt).unwrap_or_else(|| (0, 0));
next_nonce.insert(salt, (nonce + 1, expire_at));
// remove previous expiration notice
nonce_expirations.remove((block, salt));
// insert new expiration
nonce_expirations.insert((expire_at, salt), ());
```
And then the runtime can use `next_nonce` to figure out validity of the transaction, and in `on_initialize` we take all `nonce_expirations` at that particular block and prune out all the salts (since we use `DoubleMap` we can get all entries efficiently).

JoshOrndorff committed:
style cleanup on writeup

JoshOrndorff committed:
more cleanup

JoshOrndorff committed:
More writeup

bkchr commented on runtimes/frameless-runtime/src/lib.rs:
```diff
@@ -0,0 +1,301 @@
+	/// Opaque block header type.
+	pub type Header = generic::Header<BlockNumber, BlakeTwo256>;
+	/// Opaque block type.
+	pub type Block = generic::Block<Header, OpaqueExtrinsic>;
```
@JoshOrndorff you are using here the `OpaqueExtrinsic` and that works a little bit different on the node side. AFAIK it has the length prepended, which you not do.

JoshOrndorff replied:
So `OpaqueExtrinsic` is just an alias for `Vec<u8>`. And [scale docs for `Vec<_>`](https://substrate.dev/docs/en/knowledgebase/advanced/codec#vectors-lists-series-sets) do indeed say the encoding is "prefixed with a compact encoding of the number of items, followed by each item's encoding concatenated in turn".

I will be encoding a single byte which means I should prepend with the compact encoding for the number one which is `0x04` according to the same scale docs (in addition to doing the encoding myself, I notices that the number 1 is an example given.

Trying to encode my transaction this way I get the following output
```
$ curl http://localhost:9933 -H "Content-Type:application/json;charset=utf-8" -d   '{
     "jsonrpc":"2.0",
      "id":1,
      "method":"author_submitExtrinsic",
      "params": ["0x0400"]
    }'
{"jsonrpc":"2.0","error":{"code":1002,"message":"Verification Error: Execution(ApiError(\"Could not convert parameter `tx` between node and runtime: No such variant in enum FramelessTransaction\"))","data":"RuntimeApi(\"Execution(ApiError(\\\"Could not convert parameter `tx` between node and runtime: No such variant in enum FramelessTransaction\\\"))\")"},"id":1}
```

Although the error message is long, I think this is actually a step in the right direction. It is now clearly trying to decode into `FramelessTransaction` which is exactly what I want. So the remaining mysteries are:

1. Now that I know to prepend the length, how do I actually encode the `FramelessTransaction` variant?
2. Why did submitting `0x00` ever even partially work?

bkchr replied:
1. Length of 1 prepended and 1 byte added for your enum variant of the tx.

2. It is an empty transaction which probably failed in the runtime than.

JoshOrndorff replied:
> Length of 1 prepended and 1 byte added for your enum variant of the tx.

I believe that's what I've done. Length of 1 encodes as `0x04` ([citation](https://substrate.dev/docs/en/knowledgebase/advanced/codec#example-1)). My enum variant is `0x00`.

That transaction fails with the error `Verification Error: Execution(ApiError(\"Could not convert parameter `tx` between node and runtime: No such variant in enum FramelessTransaction\"`

bkchr replied:
So you had `0x0400` send to the node? (Not sure if that hex is in correct order)

JoshOrndorff replied:
Yes, submitting `0x0400` is the example quoted [above](https://github.com/substrate-developer-hub/recipes/pull/144#discussion_r439429894).

I've also tried submitting `0x0004` (which I think is backwards) and it behaves the same as `0x00`. This makes sense because the length prefix is 0 so the next byte is ignored.

gnunicorn committed:
decoding them extrinsics
